ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+Because of this we can assume that most Ruby engineers using the SDK won't be immediately familiar
+with writing async code. Because of this we have decided to make async execution opt-in, keeping
+everything synchronous by default.
+
+We have also chosen an eager execution pattern where any `async` block is started immediately and
+performed until blocked before proceeding with the execution. This is a bit easier to get a hold of
+and eliminates the risk of forgetting to actually start an async block. However this means that
+we'll need to find a solution for unhandled promise rejections, which we'll cover separately in a
+more detailed workflow proposal.
+
+We will introduce a helper method `async` to execute a block of code asynchronously:
+
+```ruby
+require 'sum_activity'
+
+class AsyncWorkflow < Temporal::Workflow
+  def execute
+    activity_promise = async { workflow.execute_activity(SumActivity, 1, 2) }
+    workflow_promise = async { workflow.execute_child_workflow(ChildWorkflow, 'parent') }
+
+    # Block execution until both promises are resolved
+    activity_result, workflow_result = async.all(activity_promise, workflow_promise).await
+
+    activity_result # => 3
+    workflow_result # => 'Hello, parent!'
+    
+    promise = async { workflow.execute_activity(SumActivity, 3, 4) }
+    
+    # individual promises can also be waited on
+    result = promise.await
+    
+    result # => 7
+  end
+end
+```
+
+A bit more involved example:
+
+```ruby
+class AnotherAsyncWorkflow < Temporal::Workflow
+  def execute
+    promise_1 = async { wait_and_sum(1, 2, delay: 15) }
+    promise_2 = async { wait_and_sum(3, 4, delay: 10) }
+
+    # The same exact method can be called in a blocking fashion
+    wait_and_sum(5, 6, delay: 5) # => 11
+
+    # Block execution until any promise is resolved
+    async.any(promise_1, promise_2).await
+
+    # Returns immediately because it will already be resolved (shorter delay)
+    promise_2.result # => 7
+
+    # This would block until the promise is resolved (longer delay)
+    promise_1.result # => 3
+  end
+
+  private
+
+  def wait_and_sum(a, b, delay:)
+    workflow.sleep(delay)
+    workflow.execute_activity(SumActivity, a, b)
+  end
+end
+```
+
+Methods `async.all` and `async.any` return promises themselves, so they can be used just like the
+promises returned by an `async {}` call.
+
+Also worth noting that `async {}` calls can be nested to model more complex business logic.
+
+
+## Activity heartbeats & cancellations
+
+This approach uses `Thread#raise` that will eject the execution out of the current context. In cases
+where this is not acceptable we will allow to either wrap a critical code in an `activity.shield`
+block (that will not raise until the block has executed) or opt-out of this default behaviour
+altogether and use `activity.cancelled?` flag to determine when an activity was cancelled.
```

Comment:
I worry about the use of `.canceled?` because when the bit gets flipped happens at unexpected times (only when heartbeating is done). I've seen a lot of confusion around this already in all SDKs, but I think the confusion would be heightened if the workflow author was able to check this bit because it would give them a false sense that it may be updated more frequently.

Reply:
While `Thread.raise` is a good way to signal cancellation, people also need to be able to know their activity is in a cancelled state. The presence of this boolean is not the same as suggesting it be the primary way to determine cancellation.

Reply:
Also worth noting that with the Rust Core the cancellation mechanism is abstracted away from heartbeating (still coupled internally, but this is not exposed to the SDK itself). So in the future cancellations might get propagated to a non-heartbeating code, so we do need to have some mechanism in place to check the state of an activity from within.